File: targets/js/sample-counter/src/sample-counter/counter.ts
import { HashCode } from "metano-runtime";

export class Counter {
  constructor(readonly count: number) { }

  static get zero(): Counter {
    return new Counter(0);
  }

  increment(): Counter {
    return new Counter(this.count + 1);
  }

  decrement(): Counter {
    return new Counter(this.count - 1);
  }

  equals(other: any): boolean {
    return other instanceof Counter && this.count === other.count;
  }

  hashCode(): number {
    const hc = new HashCode();
    hc.add(this.count);

    return hc.toHashCode();
  }

  with(overrides?: Partial<Counter>): Counter {
    return new Counter(overrides?.count ?? this.count);
  }
}

File: targets/flutter/sample_counter/lib/sample_counter/counter.dart
class Counter {
  Counter(this.count);

  final int count;

  static Counter get zero => Counter(0);

  Counter increment() => Counter(this.count + 1);

  Counter decrement() => Counter(this.count - 1);

  @override
  bool operator ==(Object other) => other is Counter && other.runtimeType == this.runtimeType && other.count == this.count;

  @override
  int get hashCode => Object.hash(this.count);

  Counter copyWith({int? count}) => Counter(count ?? this.count);
}
